### PR TITLE
Desktop: Fix #2242: Sync-Scroll for Markdown Editor and Viewer

### LINKED ISSUE #2242: When scrolling, the two sides are out of sync
## Environment

Joplin version: 1.0.175
Platform: Windows
OS specifcs: windows 10

## Steps To Reproduce

1. Create a new markdown note
2. Paste the written markdown notes
3. As you scroll through your notes, the left and right sides of the text do not correspond to each other in the title

Problem tracing:

The possible reason is that the directory was generated.
Because my notes generate a very long directory, the right side ignores that the directory takes up a lot of length when calculating the scroll interval.

### LINKED ISSUE #4827: Laggy Scrolling in Viewer
Scrolling through the Markdown viewer is quite laggy and not as smooth as it should be.

https://user-images.githubusercontent.com/78950242/114266474-31c5c480-9a14-11eb-876c-4a397f3f2f3d.mp4

## Environment

Joplin version: 1.7.11
Platform: Linux
OS specifics: Linux Mint 19.3, AppImage

## Steps to reproduce

1. Open a note. 
2. Scroll through the note with the _Viewer_.

## Describe what you expected to happen
Scrolling should be smooth and seamless without 'breaking' in the middle.

### PR DESCRIPTION
**Pull request by ken1kob**
This PR provides the synchronous scroll capability between Markdown Editor and Viewer.

It's based on [CalebJohn's branch](https://github.com/CalebJohn/joplin/tree/sync-scroll), and the following features are added.
- Bi-directional sync-scroll (viewer <-> editor)
- Performance improvements:
  - The mapping between line and scroll position is cached.
  - Efficiently observing rendering updates to keep scroll consistency
- Some cleanups

Related forum topics and github issue:
- [[Feature request] Better synchronized scrolling in desktop app](https://discourse.joplinapp.org/t/feature-request-better-synchronized-scrolling-in-desktop-app/3609)
- [Syncing Split View (Markdown/Preview)](https://discourse.joplinapp.org/t/syncing-split-view-markdown-preview/12131)
- [When scrolling, the two sides are out of sync · Issue #2242 · laurent22/joplin](https://github.com/laurent22/joplin/issues/2242)

NOTE: If you use an HD display or use zoom in/out, applying the following PR is highly recommended with this PR.
- [Desktop: Resolves #4827: Laggy Scrolling in Viewer by ken1kob · Pull Request #5496 · laurent22/joplin](https://github.com/laurent22/joplin/pull/5496)

### CODE REVIEW

**Comment by ken1kob** [comment]
## Feature Specs

### Basic Feature
- Markdown Editor and Viewer scroll synchronously.

### Detailed Features
- If you scroll Editor, Viewer is scrolled synchronously. The DOM elements corresponding to markdown lines you are looking at in Editor's viewport are located in Viewer's viewport as possible.
- The same applies in reverse. (Viewer -> Editor)
- During and after any operations, the scroll consistency between Editor and Viewer is kept as possible.
  - The followings are operations modified to adapt to sync-scroll.
	- Selecting a note
	    - When its scroll position is remembered.
		- When its rendering and image loading is delayed.
	- Editing, such as cut and paste
	- Resizing the window
	- Changing zoom factor (zoom in/out)
	- Jumping/Scrolling to a header (hash)

### Performance Requirement
- Since scrolling is a basic GUI operation, the overhead should be minimized.
  - Translating scroll positions between Editor and Viewer needs a translation map. The map should be cached, and recreated as needed.
  - Observing the changes and completion of rendering should be light-weight.

## Modified and Added Files

Short descriptions of modified and added files are here.

- MOD: NoteEditor/NoteBody/CodeMirror/CodeMirror.tsx
  - Handles scroll-related functionalities in Editor
  - Sends scroll-related messages to Viewer
  - Handles scroll-related messages from Viewer
  - Has useScrollHander
- MOD: NoteEditor/NoteBody/CodeMirror/utils/index.ts	
  - (Scroll utilities are moved to useScrollHandler.ts, because it bloated.)
- ADD: NoteEditor/NoteBody/CodeMirror/utils/useScrollHandler.ts
  - Translates scroll position from Editor to Viewer
  - Translates scroll position from Viewer to Editor
  - Handles scroll events in Editor

- MOD: NoteTextViewer.tsx	
  - Has SyncScrollMap
- MOD: note-viewer/index.html	
  - Handles scroll events in Viewer
  - Handles scroll-related messages from Editor
  - Sends scroll-related messages to Viewer
  - Updates scroll position during/after rendering
- ADD: utils/SyncScrollMap.ts	
  - Manages a line-position translation map

- ADD (renderer): MdToHtml/rules/source_map.ts
  - Embeds line-position mapping information into DOM elements.
- MOD (renderer):  MdToHtml.ts	
  - Adds source_map.ts into the renderer.

**Comment by ken1kob** [comment]
## Designs

### Abstraction View

This PR provides two functions in the abstraction view level.
- `translateScrollPercentToViewer(editorPercent)`
- `translateScrollPercentToEditor(viewerPercent)`

See the following figure. It's the abstraction view **WITHOUT** sync-scroll. A single scroll position is shared between Editor and Viewer. Contents in the two viewports of Editor and Viewer sometimes become inconsistent.

abst-wo-sync

The following two figures show the abstraction view **WITH** sync-scroll. Two separate scroll positions, e(ditor)Percent and v(iewer)Percent are used. They are translated using the above functions. That's all.

abst-with-sync

The abstraction is simple, but it explains sync-scroll's difficulty. It introduces two separate positions between Editor and Viewer. Without sync-scroll, Joplin assumed the two are the identical and have the same value in various source codes. Separating a mixture into the two is not a easy task.

### Scroll Position Translation

The following two figures explain the procedures of the translation from Editor to Viewer and from Viewer to Editor, respectively. The former is almost the same with CalebJohn's original implementation.

trans-e2v

trans-v2e

The translation is achieved using the mapping information between Editor's lines and Viewer's DOM elements' positions. The information is stored in a translation map (called `SyncScrollMap`). A target's position is estimated using linear interpolation.

### Caching Translation Map

Creating a translation map is a costly task, because it involves scanning the whole DOM document. To keep scrolling as light as possible, a created map is cached to avoid its recreations until outdated.

Since the map is cached, the binary search algorithm can be employed for querying lines or positions. It's faster than the linear search, and a query needed for one scrolling event/action is very light.

For correct operation, the map should be discarded when it becomes outdated. Simultaneously, the frequency of recreations should be minimal for users' comfort.

### Operations related to Sync-Scroll

The following existing operations are related to sync-scroll and supported in this PR.

- Editor's side:
	- scrollTo (Hash) -> sends 'scrollToHash' message to Viewer
	- scrollTo (Percent) -> sends 'setPercentScroll' message to Viewer
	- content change -> sends 'setHtml' message to Viewer
	- scroll event -> sends 'setPercentScroll' message to Viewer
	- receives 'percentScroll' message
- Viewer's side:
    - receives 'scrollToHash' message -> sends 'percentScroll' message to Editor
    - receives 'setHtml' message
    - receives 'setPercentScroll' message
    - scroll event -> sends 'percentScroll' message to Editor
    - content height change
    - completion of loading images -> send 'noteRenderComplete' message to `NoteEditor`
- Common:
	- Window resize event

### ResizeObserver

In Viewer, during a note is being rendered, its height is varying. To keep scroll consistency, observing the height and updating the scroll position is required. Since DOM elements without its counterpart window cannot rely on window.resize event, an interval timer is used to keep track of the height in the current Viewer implementation (`index.html`). Its cost is not cheap. Concretely speaking, its interval is very short (1 msec), and it sometimes miss the change of height to stop the timer. If it misses, CPU time is fully consumed for 1 sec long.

To avoid such overheads, in this PR, [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) is used. It is **standard** and an element's counterpart to window.resize event. It's more efficient and safer than using interval timer. To keep track of scrollTop and scrollHeight of a note content, MutationObserver is combined with ResizeObserver. The implementation is in `observeRendering()` and `onRendering()` of `index.html`. 

Besides, by introducing ResizeObserver, we can also know when all images are loaded. So, another interval timer,  `checkAllImageLoadedIID_` in `index.html` becomes needless.

**Comment by ken1kob** [comment]
## Items To Be Discussed

Scroll is one of the most important UI functionalities of a desktop app. Before integrating this PR, there are many concerns to be checked. About the following items, I'm still in the middle of worrying and have not made final decisions yet. They would be decided through community's discussions. I'd like to hear your advice, opinions, information, and so on.

### Item A: Sufficiency of Sync-Scroll Consistency

Is the list in the above 'Operations related to Sync-Scroll' subsection sufficient?

### Item B: Necessary and Sufficient Condition of SyncScrollMap Refresh

- Are the cases to refresh the cache of SyncScrollMap in this PR sufficient? If not sufficient, it's a bug.
- Are they minimal? If not minimal, it introduces some performance overhead.

### Item C: Functional Tests and Performance Tests

- Do you have any good notes to test this feature from the both functional and performance viewpoints?
- Is the overhead of scroll enough light-weight? How to measure it?

### Item D: Synchronization Precision

- Is the scroll synchronization enough precise?
- Is line-vs-element mapping enough? How about character-vs-element?
	- In my opinion, line-vs-element is practically sufficient.
- Using MarkdownIt, are DOM elements where mapping information are embedded appropriate?
	- The current list is in `source_map.ts`.
	- How about `img` element?
- Is the linear interpolation appropriate? Is there any other way of estimating scroll positions?

### Item E: Scroll-Past-End

Current Joplin has large (400px) scroll-past-end spaces in Editor and Viewer. Are they needed after this feature is integrated? 

See also:  [Viewer and editor panes are not in sync with Code Mirror · Issue #3341 · laurent22/joplin](https://github.com/laurent22/joplin/issues/3341)

### Item X: Others

Any other TBD items are welcome!

**Comment by ken1kob** [comment]
> If you and @CalebJohn are confident it will work well and won't introduce regression we can consider it though.

For me, it's OK.

> You might also want to check the note revision viewer because it's using the same viewer but with a slightly different (simplified) setup.

I see.

**Comment by ken1kob** [comment]
> > You might also want to check the note revision viewer because it's using the same viewer but with a slightly different (simplified) setup.

I reviewed NoteRevisionViewer. 

- The following changes of note-viewer/index.html have no impact on it.
  - sync-scroll
  - resize observer
- The sync-scroll feature of NoteTextViewer has no effect without coupling with CodeMirror.tsx.

Besides, laggy scrolling had happend in NoteRevisionViewer does not happen in the latest dev branch, since  PR #5496 is integrated.

**Comment by ken1kob** [comment]
Thank you for reviewing, @CalebJohn.

> Overall the code looks quite good! I did some testing with my "stress test" notes and couldn't find any issues. I also did some very rough performance testing and couldn't see any notable difference in CPU utilization after this PR.

That's good to hear that the PR passes your stress and load tests.

> I wonder if it would make sense to take the wonderful documentation you've prepared here and add it into the main documentation. I may be a bit biased, but I think this is a very important feature and it will be beneficial to properly document it.

About this, I have no problem.
One concern is the design might slightly change to cope with ["scroll position is not remembered"](https://discourse.joplinapp.org/t/save-cursor-position-within-the-note/43/18?u=ken1kob) problem. It might be a very deep problem.

> For Item E: The purpose of the scroll past end is for comfort of the user, having the text closer to the centre of the screen can make it easier to read. This won't need to be changed with your PR.

I see. Thanks for your explanation.

File: packages/app-desktop/gui/utils/SyncScrollMap.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		// Since creating this map is costly for each scroll event, it is cached.
+		// When some update events which outdate it such as switching a note or
+		// editing a note, it has to be invalidated (using refresh()),
+		// and a new map will be created at a next scroll event.
+		if (!doc) return null;
+		const contentElement = doc.getElementById('joplin-container-content');
+		const height = Math.max(1, contentElement.scrollHeight - contentElement.clientHeight);
+		if (this.map_) {
+			// check whether map_ is obsolete
+			if (this.map_.viewHeight == height) return this.map_;
```
Comment (laurent22) [inline packages/app-desktop/gui/utils/SyncScrollMap.ts:58]: ===

File: packages/app-desktop/gui/utils/SyncScrollMap.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	public get(doc: Document): SyncScrollMap {
+		// Returns a cached translation map between editor's scroll percenet
+		// and viewer's scroll percent. Both attributes (line and percent) of
+		// the returned map are sorted respectively.
+		// Since creating this map is costly for each scroll event, it is cached.
+		// When some update events which outdate it such as switching a note or
+		// editing a note, it has to be invalidated (using refresh()),
+		// and a new map will be created at a next scroll event.
+		if (!doc) return null;
+		const contentElement = doc.getElementById('joplin-container-content');
```
Comment (laurent22) [inline packages/app-desktop/gui/utils/SyncScrollMap.ts:54]: Please check that the element exist

File: packages/app-desktop/gui/utils/SyncScrollMap.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		const height = Math.max(1, contentElement.scrollHeight - contentElement.clientHeight);
+		if (this.map_) {
+			// check whether map_ is obsolete
+			if (this.map_.viewHeight == height) return this.map_;
+			this.map_ = null;
+		}
+		// Since getBoundingClientRect() returns a relative position,
+		// the offset of the origin is needed to get its aboslute position.
+		const offset = doc.getElementById('rendered-md').getBoundingClientRect().top;
```
Comment (laurent22) [inline packages/app-desktop/gui/utils/SyncScrollMap.ts:63]: Check that the element exist

File: packages/app-desktop/gui/note-viewer/index.html [hunk 4]
```diff
@@ -195,7 +188,66 @@
...
+					ipcProxySendToHost('syncViewerScrollWithEditor', true);
+					ipcProxySendToHost('noteRenderComplete');
+					return;
+				}
+			}
+			if (heightChanged) {
+				// When the scroll height changes, sync is needed.
+				ipcProxySendToHost('syncViewerScrollWithEditor');
+			}
+		});
```
Comment (laurent22) [inline packages/app-desktop/gui/note-viewer/index.html:250]: Do you know if there's a performance impact on very large notes?

Also the notes sometimes are giant blobs of HTML, when it's imported from the web clipper or Evernote. Could you check that it works with these too (in particular that it doesn't freeze)?

File: packages/renderer/MdToHtml/rules/source_map.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+			blockquote_open: 0,
+			table_open: 0,
+			code_block: 0,
+			hr: 0,
+			html_block: 0,
+			list_item_open: 99, // this will stop matching if a list goes more than 99 indents deep
+			math_block: 0,
+		};
+
+		for (const [key, allowed_level] of Object.entries(allowed_levels)) {
```
Comment (laurent22) [inline packages/renderer/MdToHtml/rules/source_map.ts:19]: allowedLevel

File: packages/renderer/MdToHtml/rules/source_map.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+export default {
+	plugin: (markdownIt: any, params: any) => {
+
+		if (!params.mapsToLine) return;
+
+		const allowed_levels = {
```
Comment (laurent22) [inline packages/renderer/MdToHtml/rules/source_map.ts:6]: allowedLevels (camel case everywhere, as much as possible)

File: packages/renderer/MdToHtml/rules/source_map.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+			table_open: 0,
+			code_block: 0,
+			hr: 0,
+			html_block: 0,
+			list_item_open: 99, // this will stop matching if a list goes more than 99 indents deep
+			math_block: 0,
+		};
+
+		for (const [key, allowed_level] of Object.entries(allowed_levels)) {
+			const precedent_rule = markdownIt.renderer.rules[key];
```
Comment (laurent22) [inline packages/renderer/MdToHtml/rules/source_map.ts:20]: precedentRule

**Comment by laurent22** [comment]
Thanks for the update @ken1kob, I've just added a few more comments. As Caleb suggested, could you also please add your documentation under readme/spec? And then a link to it in README.md, in the Table of contents. 

And that would be great if you could also keep your images - you can add them under Assets/WebsiteAssets/images/SPEC_NAME, then link to it from the Markdown document using ``

File: packages/app-desktop/gui/note-viewer/index.html (hunk 4 above)
Comment (tessus) [reply packages/app-desktop/gui/note-viewer/index.html:250]: I also tested perf. The longest note was 61297 words with 3000 lines and I didn't notice a thing.

File: packages/app-desktop/gui/utils/SyncScrollMap.ts (hunk 1 above)
Comment (ken1kob) [reply packages/app-desktop/gui/utils/SyncScrollMap.ts]: Finished. 50bb62f

File: packages/app-desktop/gui/utils/SyncScrollMap.ts (hunk 2 above)
Comment (ken1kob) [reply packages/app-desktop/gui/utils/SyncScrollMap.ts:54]: Finished. 50bb62f

File: packages/app-desktop/gui/utils/SyncScrollMap.ts (hunk 3 above)
Comment (ken1kob) [reply packages/app-desktop/gui/utils/SyncScrollMap.ts:64]: Finished. 50bb62f

File: packages/renderer/MdToHtml/rules/source_map.ts (hunk 5 above)
Comment (ken1kob) [reply packages/renderer/MdToHtml/rules/source_map.ts]: Finished. 50bb62f

File: packages/renderer/MdToHtml/rules/source_map.ts (hunk 6 above)
Comment (ken1kob) [reply packages/renderer/MdToHtml/rules/source_map.ts]: Finished. 50bb62f

File: packages/renderer/MdToHtml/rules/source_map.ts (hunk 7 above)
Comment (ken1kob) [reply packages/renderer/MdToHtml/rules/source_map.ts]: Finished. 50bb62f

File: packages/app-desktop/gui/note-viewer/index.html (hunk 4 above)
Comment (ken1kob) [reply packages/app-desktop/gui/note-viewer/index.html:250]: I also tested a web-clipped page from [Long pages - Wikipedia](https://en.wikipedia.org/wiki/Special:LongPages). It consist of 2k lines and 60k words. There is no problem found.

**Comment by ken1kob** [comment]
> As Caleb suggested, could you also please add your documentation under readme/spec? And then a link to it in README.md, in the Table of contents.

I see. I'll add the document in an independent PR after I solve ["scroll position is not remembered"](https://discourse.joplinapp.org/t/save-cursor-position-within-the-note/43/18?u=ken1kob) problem. Because I anticipate the design will slightly need to be changed to solve it.

**Comment by laurent22** [comment]
Sure, a separate pull request would work too.

So I think we're good now, thanks a lot for working on this @ken1kob, that's a pretty impressive change! And thanks @CalebJohn and @tessus for checking and reviewing!

**Comment by CalebJohn** [comment]
Thanks @ken1kob @laurent22  and @tessus 

This was a big change and one I've been looking forward too for a long time. 
Particularly big thanks to Kenichi for your persistence!

**Comment by ken1kob** [comment]
Thanks for your collaborations, @laurent22 @CalebJohn and @tessus. It's the first time for me such a big feature is integrated to Joplin. It's my pleasure, and the collaborating development was a lot of fun.

**Comment by antoniopaolini** [comment]
Hello, 
I have a note (at the moment a single note, I should look further) that go crazy with the new Sync -Scroll.
It' is a table, a very long table, and if I add some empty lines (I use to add lines and then popolate it later) the note is TOTALLY out of sync.
This happens with 
```
Joplin 2.6.2 (prod, win32)

ID del Client: 6c64c797f00b410c98d19c8697f4355e
Sinc. versione: 3
Profilo versione: 41
Portachiavi supportato: No

Revisione: 5a82f5d
```
With the previous 2.5.12 the behavior was "normal".

How (and where) I can send this note? In what format is preferable (jex? md?)

Thanks.

**Comment by ken1kob** [comment]
Hi,  @antoniopaolini. 

Thanks for your report. But, could you post a bug report as a new issue? Since this is a closed PR, a new issue is preferable to be easily tracked for everyone.

> How (and where) can I send this note? What format is preferrable (jex? md? raw?)

If it is not so large, attaching md to the post would be handy.

**Comment by antoniopaolini** [comment]
Ok, thanks.
I done it (#5808)

**Comment by laurent22** [comment]
It seems to work better when disabling inline images in the Rich Markdown plugin. So it might be that the sync-scroll algorithm is making assumptions about line heights, which are no longer true when images are being displayed inline.

**Comment by laurent22** [comment]
@ken1kob, alternatively, even if we don't revert the change, do you think it's possible to temporarily disable it (so that it scrolls like it did before)? It's blocking the 2.6 release so your recommendation on what's best to do would be appreciated.

**Comment by ken1kob** [comment]
@laurent22 and @CalebJohn ,

I inspected the behavior using 'Ultimate Note.jex'. It is the mixture of three symptoms.
- (A) Continous scroll events after long scrolling. The intervals are 3000 msec , and several scroll events are emitted for each interval. 
- (B) Drift of a scroll position after long scrolling
- (C) Jump of a scroll position after long scrolling and note switching in Editor layout mode

Causes:
- (A) Scroll events are emitted from the interval timer handler, `ImageHandlers.refreshAllWidgets` in RichMarkdownEditor. After a long scrolling, the handler continuously calls `cm.scrollTo()`, until something (ex. typing a character) happens.
  - Ref: https://github.com/CalebJohn/joplin-rich-markdown/blob/f54849ee9b7b89a5d0aae019bec8a14dad344bd6/src/richMarkdown.ts#L49
   ```js
    if (settings.inlineImages && settings.scanImgChangePeriod > 0) {
        this.state.richMarkdown.scanImgInterval = setInterval(ImageHandlers.refreshAllWidgets, settings.scanImgChangePeriod * 1000, this);
    }
  ```
  - I still don't know why the handler continuously calls `cm.scrollTo()`.
  
- (B) In the handler, the old position which was given when the handler closure was created is kept. After a maximum of 3000 msec,  `cm.scrollTo()` is invoked and  the old position is remembered. Since it is incompatible to sync-scroll position translation, it results in a drift.

- (C) When Viewer is not visible, images in a note are not loaded in Viewer. So, a 'noteRenderComplete' message is not sent to Editor and `scrollWhenReady` is not consumed. When a character is typed in Editor, a 'noteRenderComplete' message is sent to Editor even if Viewer is not visible. In this situation, unconsumed `scrollWhenReady` is used, and the recorded scroll position (in most cases, the top of the note) is remembered. It is 'jump of scroll position'.
	- In pre sync scroll, noteRenderComplete is expired in 1000msec. So, this symptom does not appear.

Symptoms in the new sync scroll (#5826):
- (A) Almost harmless. Because the new sync scroll ignores the same scroll events in a sequence.
- (B) It happens.
- (C) It does not happen. Because 'noteRenderComplete' does not wait the completion of loading images.

Treatments in the current sync scroll (#5512):
- (A) Possible by porting relevant codes from #5826.
- (B) Under investigation
- (C) It can be mitigated by making expiration time for 'noteRenderComplete'

Treatments in the new sync scroll:
- (A) Not needed
- (B) There are two approaches:
	- (B-1) Modifying RichMarkdownEditor: Removing `cm.scrollTo()` in refreshAllWidgets() in RichMarkdownEditor. In the new sync scroll, it's OK, because scroll positions are restored when Editor is resized.
	- (B-2) Modifying Joplin: Under investigation
- (C) Not needed

**Comment by ken1kob** [comment]
>  It's blocking the 2.6 release so your recommendation on what's best to do would be appreciated.

My recommendation is integrating #5826. It is more robust than #5512, and the remaining symptom (B) seems to be not urgent.

**Comment by laurent22** [comment]
That's great, thanks for looking into it and for the detailed explanation @ken1kob. As I understand the Rich Markdown plugin restores the scroll position after images are loaded, and indeed that's probably interfering with your new scroll logic. So you're saying we can probably remove this scroll hack from the plugin as it's no longer needed?

**Comment by ken1kob** [comment]
> So you're saying we can probably remove this scroll hack from the plugin as it's no longer needed?

Yes. But only if #5826 is integrated. #5512 does not have restoring after resizing in Editor.

**Comment by laurent22** [comment]
Ok thanks @ken1kob, I've now merged the pull request.

@CalebJohn, I've tried to remove the scrollTo line in the Rich Markdown plugin and indeed fixes the issue as Kenichi mentioned:

https://github.com/CalebJohn/joplin-rich-markdown/blob/da7454d65e07540aa8f92abf7dc2a6143dcd110a/src/images.ts#L262

It does fix the issue and I didn't notice any regression, although I'm not that familiar with the plugin. Can you think of any issue that removing this line could cause?

**Comment by CalebJohn** [comment]
Thanks @ken1kob for tracking down that issue
Indeed I knew it wasn't the most elegant bit of code, but, as mentioned, it was necessary to keep the scrolling somewhat aligned as images load in.
The reason it calls it every 3000ms is because of the image monitoring. It's a feature (that can be disabled) that checks for on disk changes of images and then re-loads them. 

> @CalebJohn, I've tried to remove the scrollTo line in the Rich Markdown plugin and indeed fixes the issue as Kenichi mentioned:
> It does fix the issue and I didn't notice any regression, although I'm not that familiar with the plugin. Can you think of any issue that removing this line could cause?

No, there shouldn't be any issue (although I haven't tested that with the latest sync scroll changes). Unfortunately, I'm back in my hometown for the holidays and don't have my dev machine with me. That said, I think I can get set up well enough to remove this from the plugin and push out a new release (I just won't have a full Joplin dev environment).

**Comment by matypist** [comment]
You can't imagine how long I've wanted something like this, thank you so much!

**Pull request merged**